### PR TITLE
[python] Support default-coord reads of corner-written dense ND arrays

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** In support of PR #1768 with issue #1769.

Needles to thread for `update_uns`:

The existing `tiledbsoma.io.update_obs`, `tiledbsoma.io.update_var`, `tiledbsoma.io.add_X_layer` methods all take as argument a `tiledbsoma.Experiment` which is already open for writing. Goal: do the same for `tiledbsoma.io.update_uns`.

Common use-case:
* Ingest data from AnnData to SOMA
* Outgest to AnnData; run ScanPy, curation, etc
* Write the updates back to SOMA

Technical factors:
* Whereas all the `X`, `obsm`, `varm`, `obsp`, `varp` are `SparseNDArray`, `uns` can contain `DenseNDArray` which are ingested from `numpy.ndarray`
* If the after-uns has `numpy.ndarray` that the before-uns didn't (new data on update), there is no problem
* If the after-uns has `numpy.ndarray` of the same dimensions as the before-uns, there is no problem
* Schema evolution does not support modification of dense domains
* We can try delete-and-re-add:
  * `tiledb.Array.delete_array` of the array accompanied by tiledb-group operations to delete the key/URI pair and re-add the key/URI pair
  * Note that for local-disk and S3, the new URI will be the same as the old URI; for cloud the new and old URIs will always be different URIs
  * Problem with that: we cannot delete-and-re-add at the same timestamp
  * If we do `del collection[key]` and `collection[key] = new_dense_nd_array` then we get `tiledbsoma._exception.SOMAError: replacing key 'a' is unsupported` which is a check that must remain (see links below)
* Also we cannot do: del collection[key], close the collection, and re-open at the same timestamp because of
  * https://github.com/single-cell-data/TileDB-SOMA/blob/1.5.0rc2/apis/python/src/tiledbsoma/_collection.py#L567-L570
  * https://github.com/single-cell-data/TileDB-SOMA/pull/957
  * https://github.com/single-cell-data/TileDB-SOMA/issues/919
* We could have `update_uns` do two opens-for-write: one to remove any shape-changing dense arrays, and one shifted by a millisecond to do the rest of the work
  * However this would mean `tiledbsoma.io.update_uns` would take a URI as first argument, different from all the other updaters in `tiledbsoma.io` which take an `Experiment` opened for write as first argument

Good feedback from @isaiah and @seth: simply have `tiledbsoma.io`'s `uns`-ingestion logic write bigger-than-necessary arrays for dense ND arrays -- as we already do for sparse.

This is compliant with the [SOMA specification](https://github.com/single-cell-data/SOMA/blob/main/abstract_specification.md#indexing-and-slicing).

**Changes:** As above.

**Notes for Reviewer:**

### CODE REVIEW

File: apis/python/src/tiledbsoma/_dense_nd_array.py [hunk 1]
```diff
@@ -71,6 +72,18 @@ class DenseNDArray(NDArray, somacore.DenseNDArray):
...
+        Retrieves the non-empty domain for each dimension, namely the smallest
+        and largest indices in each dimension for which the sparse array has
+        data occupied.  This is nominally the same as the domain used at
+        creation time, but if for example only a corner of the available domain
+        has actually had data written, this function will return a tighter
+        range.
+        """
+        with tiledb.open(self.uri, ctx=self.context.tiledb_ctx) as A:
+            return A.nonempty_domain()  # type: ignore
+
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/_dense_nd_array.py:86]: What I recommend instead is to remove this, and add a `TileDBArray._tiledb_nonempty_domain()` function into [_tiledb_array.py](https://github.com/single-cell-data/TileDB-SOMA/blob/main/apis/python/src/tiledbsoma/_tiledb_array.py#L45). This uses the already opened handle and does not introduce a new `DenseNDArray.non_empty_domain`.

On this line, call `ned = self._handle._tiledb_nonempty_domain()` which then just keep getting the NED internal.
https://github.com/single-cell-data/TileDB-SOMA/pull/1810/files#diff-1c8b39ff256e708f483b5e9ebdb3c0dd07f14a78cd487cbf1e8d7c49373d9997R137

Comment (nguyenv) [reply apis/python/src/tiledbsoma/_dense_nd_array.py:86]: Also minor nit, but I think we favor naming as `nonempty_domain` over `non_empty_domain`.

Comment (johnkerl) [reply apis/python/src/tiledbsoma/_dense_nd_array.py:86]: @nguyenv it's already released for sparse (i'm adding it for dense) and i don't want a backward-incompatible name change

thanks though!

Comment (nguyenv) [reply apis/python/src/tiledbsoma/_dense_nd_array.py:86]: OK nevermind about the name change. But I think the fact that both `SparseNDArray` and `DenseNDArray` use this is even more reason to add this into `TileDBArray` and then in both classes use

```
    def non_empty_domain(self) -> Tuple[Tuple[int, int], ...]:
        """
        Retrieves the non-empty domain for each dimension, namely the smallest and
        largest indices in each dimension for which the sparse array has data occupied.
        This is nominally the same as ``used_shape``, but if for example the
        leading/trailing rows/columns of the sparse array are entirely unoccupied, this
        function will return a tighter range.
        """
        return self._handle._tiledb_nonempty_domain()
```

Comment (johnkerl) [reply apis/python/src/tiledbsoma/_dense_nd_array.py]: just pushed a commit w/ that change! :)

Comment (johnkerl) [reply apis/python/src/tiledbsoma/_dense_nd_array.py]:
```
>>> exp.obs._handle. <---- tab-complete
exp.obs._handle.close(       exp.obs._handle.metadata     exp.obs._handle.reader       exp.obs._handle.uri
exp.obs._handle.closed       exp.obs._handle.mode         exp.obs._handle.schema       exp.obs._handle.writer
exp.obs._handle.context      exp.obs._handle.open(        exp.obs._handle.timestamp_ms

>>> type(exp.obs._handle)
<class 'tiledbsoma._tdb_handles.ArrayWrapper'>

>>> exp.obs._handle._
exp.obs._handle._abc_impl          exp.obs._handle._flush_hack(       exp.obs._handle._opener(
exp.obs._handle._check_open(       exp.obs._handle._handle
exp.obs._handle._do_initial_reads( exp.obs._handle._is_protocol
```

so we would need a follow-on once we've gotten `nonempty_domain` wired in from C++

**Review by nguyenv — APPROVED** [review]
Thanks!

**Pull request merged**